### PR TITLE
feat(optimizer): `Adagrad` will use `device` when `capturable` - True always when compiling with dynamo

### LINKED ISSUE #107006: Apply fusion more aggressively in NAdam and Adagrad compilation
### 🚀 The feature, motivation and pitch

Currently NAdam compiles to > 5 kernels and Adagrad does as well. See https://github.com/pytorch/pytorch/blob/4df84c3b4d93ea53c9ca169f32a0dc4619bee797/test/inductor/test_compiled_optimizers.py#L4

for current kernel counts. Ideally these should fully fuse, but likely there are issues due to the presence of mutation. A good place to start is https://github.com/pytorch/pytorch/blob/4df84c3b4d93ea53c9ca169f32a0dc4619bee797/torch/_inductor/scheduler.py#L4

Specifically https://github.com/pytorch/pytorch/blob/4df84c3b4d93ea53c9ca169f32a0dc4619bee797/torch/_inductor/scheduler.py#L644

Ideally the rules should be able to be modified to soundly allow these to fuse fully. 

Contact @mlazos for more details

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @wconstab @Xia-Weiwen @ngimel

### PR DESCRIPTION
**Pull request by jon-chuang**
Partial fix: https://github.com/pytorch/pytorch/issues/107006

CC: @mlazos as issue creator

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/optim/adagrad.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def __init__(
         for group in self.param_groups:
             for p in group["params"]:
                 state = self.state[p]
-                state["step"] = torch.tensor(0.0)
+                state["step"] = torch.tensor(0.0, device=p.device)
```
Comment (jon-chuang) [inline torch/optim/adagrad.py:53]: This will actually slowdown eager code, as eager code would prefer this to be in CPU. 

But it seems necessary to overcome the device mismatch in the buffers during fusion scheduling. Nevertheless, it is odd that e.g. `Adam` does not display this requirement - even if its `step` tensor is on CPU, all of its input buffers end up being on GPU.

Comment (jon-chuang) [reply torch/optim/adagrad.py:53]: That was in theory - it seems that in practice there is no difference:

image

Comment (jon-chuang) [reply torch/optim/adagrad.py:53]: A comparison for `torch.compile` would also be interesting - to understand the practical speedup of having fewer kernels.

File: torch/optim/adagrad.pyi [hunk 2]
```diff
@@ -9,4 +9,5 @@ class Adagrad(Optimizer):
         weight_decay: float = ...,
         initial_accumulator_value: float = ...,
         eps: float = ...,
+        capturable: bool = False,
```
Comment (jon-chuang) [inline torch/optim/adagrad.pyi:12]: nit: not sure if this is necessary. Not sure what the `pyi` is used for tbh.

File: torch/optim/adagrad.py [hunk 3]
```diff
@@ -44,28 +45,18 @@ def __init__(
             foreach=foreach,
             maximize=maximize,
             differentiable=differentiable,
+            capturable=capturable,
         )
         super().__init__(params, defaults)
 
-        for group in self.param_groups:
```
Comment (jon-chuang) [inline torch/optim/adagrad.py:50]: Thanks to this change, a previous error in test/nn/test_lazy_modules.py for lazy initialization is now avoided.

File: test/inductor/test_compiled_optimizers.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def tearDown(self):
     test_rprop_recompile = make_recompile_test(Rprop, kernel_count=1, lr=0.01)
     test_rmsprop_recompile = make_recompile_test(RMSprop, kernel_count=1, lr=0.01)
     test_adadelta_recompile = make_recompile_test(Adadelta, kernel_count=1, lr=0.01)
-    test_adagrad_recompile = make_recompile_test(Adagrad, kernel_count=5, lr=0.01)
+    test_adagrad_recompile = make_recompile_test(Adagrad, kernel_count=2, lr=0.01)
```
Comment (jon-chuang) [inline test/inductor/test_compiled_optimizers.py:189]: I think adding more configs like to adamax is a good idea.

File: test/nn/test_lazy_modules.py [hunk 5]
```diff
@@ -605,10 +605,6 @@ def run_step(module, optim):
             module.register_parameter('test_param', UninitializedParameter())
             if optim_cls is torch.optim.SGD:
                 optim = optim_cls(module.parameters(), lr=0.0)
-            elif optim_cls is torch.optim.Adagrad:
```
Comment (jon-chuang) [inline test/nn/test_lazy_modules.py:608]: Note to reviewer:

As mentioned below, as we now lazily initialize step, this previous error in test/nn/test_lazy_modules.py for lazy initialization is now avoided.

File: torch/optim/adagrad.pyi (hunk 2 above)
Comment (jon-chuang) [reply torch/optim/adagrad.pyi:12]: anyway, small issue, it's better to have more type info imo

**Comment by jon-chuang** [comment]
Ping @janeyx99 @mlazos 

With the other issues out of the way, I hope we're able to tackle the original investigation into improving fusion of optimizers!

File: torch/optim/adagrad.py [hunk 6]
```diff
@@ -84,6 +75,23 @@ def share_memory(self):
     def _init_group(self, group, params_with_grad, grads, state_sums, state_steps):
         has_sparse_grad = False
         for p in group["params"]:
+            state = self.state[p]
+            if "step" not in state:
+                state["step"] = (
+                    torch.zeros((), dtype=torch.float, device=p.device)
```
Comment (janeyx99) [inline torch/optim/adagrad.py:81]: let’s use p.new_zeros to maintain subclass information of p

**Review by janeyx99 — COMMENTED** [review]
Thanks for adding the capturable path. Adding capturable is heftier as we should support for single tensor as well as adding cuda graphs testing. Feel free to follow the example in https://github.com/pytorch/pytorch/pull/106615 for what is expected. It might make sense to open general capturable in another PR and have this one be the compiler stuff built on top of it.
